File: ts/packages/over-memory/src/bus.ts
import type { Dispatcher } from '@clamator/protocol';

export class MemoryBus {
  private dispatchers = new Map<string, Dispatcher>();

  register(service: string, dispatch: Dispatcher): void {
    if (this.dispatchers.has(service))
      throw new Error(`service "${service}" already registered on this bus`);
    this.dispatchers.set(service, dispatch);
  }

  unregister(service: string): void {
    this.dispatchers.delete(service);
  }

  lookup(service: string): Dispatcher | undefined {
    return this.dispatchers.get(service);
  }
}

File: py/packages/over-memory/src/clamator_over_memory/bus.py
from __future__ import annotations

from clamator_protocol import Dispatcher


class MemoryBus:
    def __init__(self) -> None:
        self._dispatchers: dict[str, Dispatcher] = {}

    def register(self, service: str, dispatch: Dispatcher) -> None:
        if service in self._dispatchers:
            raise ValueError(f'service "{service}" already registered on this bus')
        self._dispatchers[service] = dispatch

    def unregister(self, service: str) -> None:
        self._dispatchers.pop(service, None)

    def lookup(self, service: str) -> Dispatcher | None:
        return self._dispatchers.get(service)
